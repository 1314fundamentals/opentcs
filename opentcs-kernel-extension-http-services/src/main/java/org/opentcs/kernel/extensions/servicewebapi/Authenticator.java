// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
