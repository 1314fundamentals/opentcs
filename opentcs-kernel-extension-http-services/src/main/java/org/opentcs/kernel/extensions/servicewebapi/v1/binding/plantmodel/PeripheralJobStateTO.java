// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel;

public enum PeripheralJobStateTO {
  // CHECKSTYLE:OFF
  TO_BE_PROCESSED,
  BEING_PROCESSED,
  FINISHED,
  FAILED
  // CHECKSTYLE:ON
}
