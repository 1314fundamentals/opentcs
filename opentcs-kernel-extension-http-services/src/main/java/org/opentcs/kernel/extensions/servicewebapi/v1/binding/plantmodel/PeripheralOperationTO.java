// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.annotation.Nonnull;

/**
 */
public class PeripheralOperationTO {

  private String operation;
  private String locationName;
  private ExecutionTrigger executionTrigger = ExecutionTrigger.AFTER_ALLOCATION;
  private boolean completionRequired;

  @JsonCreator
  public PeripheralOperationTO(
      @Nonnull
      @JsonProperty(value = "operation", required = true)
      String operation,
      @Nonnull
      @JsonProperty(value = "locationName", required = true)
      String locationName
  ) {
    this.operation = requireNonNull(operation, "operation");
    this.locationName = requireNonNull(locationName, "locationName");
  }

  @Nonnull
  public String getOperation() {
    return operation;
  }

  public PeripheralOperationTO setOperation(
      @Nonnull
      String operation
  ) {
    this.operation = requireNonNull(operation, "operation");
    return this;
  }

  @Nonnull
  public String getLocationName() {
    return locationName;
  }

  public PeripheralOperationTO setLocationName(
      @Nonnull
      String locationName
  ) {
    this.locationName = requireNonNull(locationName, "locationName");
    return this;
  }

  @Nonnull
  public ExecutionTrigger getExecutionTrigger() {
    return executionTrigger;
  }

  public PeripheralOperationTO setExecutionTrigger(
      @Nonnull
      ExecutionTrigger executionTrigger
  ) {
    this.executionTrigger = requireNonNull(executionTrigger, "executionTrigger");
    return this;
  }

  public boolean isCompletionRequired() {
    return completionRequired;
  }

  public PeripheralOperationTO setCompletionRequired(boolean completionRequired) {
    this.completionRequired = completionRequired;
    return this;
  }

  // CHECKSTYLE:OFF
  public enum ExecutionTrigger {

    IMMEDIATE,
    AFTER_ALLOCATION,
    AFTER_MOVEMENT;
  }
  // CHECKSTYLE:ON
}
