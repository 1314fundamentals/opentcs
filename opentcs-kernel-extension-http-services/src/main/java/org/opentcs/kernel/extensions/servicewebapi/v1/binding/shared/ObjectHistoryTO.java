// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import java.time.Instant;
import java.util.List;

/**
 */
public class ObjectHistoryTO {
  private List<ObjectHistoryEntryTO> entries;

  public ObjectHistoryTO(
      List<ObjectHistoryEntryTO> entries
  ) {
    this.entries = requireNonNull(entries, "entries");
  }

  public List<ObjectHistoryEntryTO> getEntries() {
    return entries;
  }

  public ObjectHistoryTO setEntries(List<ObjectHistoryEntryTO> entries) {
    this.entries = entries;
    return this;
  }

  /**
   */
  public static class ObjectHistoryEntryTO {
    private Instant timestamp;
    private String eventCode;
    private List<String> supplements;

    public ObjectHistoryEntryTO(
        Instant timeStamp,
        String eventCode,
        List<String> supplements
    ) {
      this.timestamp = requireNonNull(timeStamp, "timeStamp");
      this.eventCode = requireNonNull(eventCode, "eventCode");
      this.supplements = requireNonNull(supplements, "supplements");
    }

    public Instant getTimestamp() {
      return timestamp;
    }

    public ObjectHistoryEntryTO setTimestamp(Instant timestamp) {
      this.timestamp = timestamp;
      return this;
    }

    public String getEventCode() {
      return eventCode;
    }

    public ObjectHistoryEntryTO setEventCode(String eventCode) {
      this.eventCode = eventCode;
      return this;
    }

    public List<String> getSupplements() {
      return supplements;
    }

    public ObjectHistoryEntryTO setSupplements(List<String> supplements) {
      this.supplements = supplements;
      return this;
    }
  }
}
