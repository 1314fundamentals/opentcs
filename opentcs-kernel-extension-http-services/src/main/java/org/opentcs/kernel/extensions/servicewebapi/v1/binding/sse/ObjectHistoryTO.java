// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.opentcs.data.ObjectHistory;

/**
 * A transfer object (to be used with the SSE API) representing an {@link ObjectHistory} instance.
 */
// CHECKSTYLE:OFF
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
public class ObjectHistoryTO {

  private List<ObjectHistoryEntryTO> entries;

  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class ObjectHistoryEntryTO {

    private Instant timestamp;
    private String eventCode;
    private List<String> supplements;
  }
}
// CHECKSTYLE:ON
