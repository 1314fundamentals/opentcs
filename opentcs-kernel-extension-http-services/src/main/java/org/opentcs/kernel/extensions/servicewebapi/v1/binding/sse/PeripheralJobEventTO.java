// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.Map;
import javax.annotation.Nullable;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.opentcs.data.TCSObjectEvent;

/**
 * A transfer object (to be used with the SSE API) representing a {@link TCSObjectEvent} for a
 * peripheral job.
 */
public class PeripheralJobEventTO {

  private final PeripheralJobTO currentObjectState;
  private final PeripheralJobTO previousObjectState;

  public PeripheralJobEventTO(
      @Nullable
      PeripheralJobTO currentObjectState,
      @Nullable
      PeripheralJobTO previousObjectState
  ) {
    this.currentObjectState = currentObjectState;
    this.previousObjectState = previousObjectState;
  }

  public PeripheralJobTO getCurrentObjectState() {
    return currentObjectState;
  }

  public PeripheralJobTO getPreviousObjectState() {
    return previousObjectState;
  }

  // CHECKSTYLE:OFF
  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class PeripheralJobTO {

    private String name;
    private Map<String, String> properties;
    private ObjectHistoryTO history;
    private String reservationToken;
    private String relatedVehicle;
    private String relatedTransportOrder;
    private PeripheralOperationTO peripheralOperation;
    private StateTO state;
    private Instant creationTime;
    private Instant finishedTime;

    @NoArgsConstructor
    @Getter
    @Setter
    @Accessors(chain = true)
    public static class PeripheralOperationTO {

      private String location;
      private String operation;
      private ExecutionTriggerTO executionTrigger;
      private boolean completionRequired;

      public enum ExecutionTriggerTO {
        IMMEDIATE,
        AFTER_ALLOCATION,
        AFTER_MOVEMENT
      }
    }

    public enum StateTO {
      TO_BE_PROCESSED,
      BEING_PROCESSED,
      FINISHED,
      FAILED
    }
    // CHECKSTYLE:ON
  }
}
