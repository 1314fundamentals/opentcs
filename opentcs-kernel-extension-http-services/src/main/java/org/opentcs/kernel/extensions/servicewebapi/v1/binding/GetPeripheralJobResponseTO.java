// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import java.time.Instant;
import java.util.List;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel.PeripheralJobStateTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.PeripheralOperationDescription;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * The current state of a peripheral job.
 */
public class GetPeripheralJobResponseTO {

  private String name;

  private String reservationToken;

  private String relatedVehicle;

  private String relatedTransportOrder;

  private PeripheralOperationDescription peripheralOperation;

  private PeripheralJobStateTO state;

  private Instant creationTime;

  private Instant finishedTime;

  private List<Property> properties;

  public GetPeripheralJobResponseTO() {
  }

  public String getName() {
    return name;
  }

  public GetPeripheralJobResponseTO setName(String name) {
    this.name = name;
    return this;
  }

  public String getReservationToken() {
    return reservationToken;
  }

  public GetPeripheralJobResponseTO setReservationToken(String reservationToken) {
    this.reservationToken = reservationToken;
    return this;
  }

  public String getRelatedVehicle() {
    return relatedVehicle;
  }

  public GetPeripheralJobResponseTO setRelatedVehicle(String relatedVehicle) {
    this.relatedVehicle = relatedVehicle;
    return this;
  }

  public String getRelatedTransportOrder() {
    return relatedTransportOrder;
  }

  public GetPeripheralJobResponseTO setRelatedTransportOrder(String relatedTransportOrder) {
    this.relatedTransportOrder = relatedTransportOrder;
    return this;
  }

  public PeripheralOperationDescription getPeripheralOperation() {
    return peripheralOperation;
  }

  public GetPeripheralJobResponseTO setPeripheralOperation(
      PeripheralOperationDescription peripheralOperation
  ) {
    this.peripheralOperation = peripheralOperation;
    return this;
  }

  public PeripheralJobStateTO getState() {
    return state;
  }

  public GetPeripheralJobResponseTO setState(PeripheralJobStateTO state) {
    this.state = state;
    return this;
  }

  public Instant getCreationTime() {
    return creationTime;
  }

  public GetPeripheralJobResponseTO setCreationTime(Instant creationTime) {
    this.creationTime = creationTime;
    return this;
  }

  public Instant getFinishedTime() {
    return finishedTime;
  }

  public GetPeripheralJobResponseTO setFinishedTime(Instant finishedTime) {
    this.finishedTime = finishedTime;
    return this;
  }

  public List<Property> getProperties() {
    return properties;
  }

  public GetPeripheralJobResponseTO setProperties(List<Property> properties) {
    this.properties = properties;
    return this;
  }
}
