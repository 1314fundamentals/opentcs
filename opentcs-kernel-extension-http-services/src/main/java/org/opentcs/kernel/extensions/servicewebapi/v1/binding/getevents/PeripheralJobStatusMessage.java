// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents;

import java.time.Instant;
import java.util.List;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel.PeripheralJobStateTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.PeripheralOperationDescription;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * A status message containing information about a peripheral job.
 */
public class PeripheralJobStatusMessage
    extends
      StatusMessage {

  private String name;

  private String reservationToken;

  private String relatedVehicle;

  private String relatedTransportOrder;

  private PeripheralOperationDescription peripheralOperation;

  private PeripheralJobStateTO state;

  private Instant creationTime;

  private Instant finishedTime;

  private List<Property> properties;

  /**
   * Creates a new instance.
   */
  public PeripheralJobStatusMessage() {
  }

  @Override
  public PeripheralJobStatusMessage setSequenceNumber(long sequenceNumber) {
    return (PeripheralJobStatusMessage) super.setSequenceNumber(sequenceNumber);
  }

  @Override
  public PeripheralJobStatusMessage setCreationTimeStamp(Instant creationTimeStamp) {
    return (PeripheralJobStatusMessage) super.setCreationTimeStamp(creationTimeStamp);
  }

  public String getName() {
    return name;
  }

  public PeripheralJobStatusMessage setName(String name) {
    this.name = name;
    return this;
  }

  public String getReservationToken() {
    return reservationToken;
  }

  public PeripheralJobStatusMessage setReservationToken(String reservationToken) {
    this.reservationToken = reservationToken;
    return this;
  }

  public String getRelatedVehicle() {
    return relatedVehicle;
  }

  public PeripheralJobStatusMessage setRelatedVehicle(String relatedVehicle) {
    this.relatedVehicle = relatedVehicle;
    return this;
  }

  public String getRelatedTransportOrder() {
    return relatedTransportOrder;
  }

  public PeripheralJobStatusMessage setRelatedTransportOrder(String relatedTransportOrder) {
    this.relatedTransportOrder = relatedTransportOrder;
    return this;
  }

  public PeripheralOperationDescription getPeripheralOperation() {
    return peripheralOperation;
  }

  public PeripheralJobStatusMessage setPeripheralOperation(
      PeripheralOperationDescription peripheralOperation
  ) {
    this.peripheralOperation = peripheralOperation;
    return this;
  }

  public PeripheralJobStateTO getState() {
    return state;
  }

  public PeripheralJobStatusMessage setState(PeripheralJobStateTO state) {
    this.state = state;
    return this;
  }

  public Instant getCreationTime() {
    return creationTime;
  }

  public PeripheralJobStatusMessage setCreationTime(Instant creationTime) {
    this.creationTime = creationTime;
    return this;
  }

  public Instant getFinishedTime() {
    return finishedTime;
  }

  public PeripheralJobStatusMessage setFinishedTime(Instant finishedTime) {
    this.finishedTime = finishedTime;
    return this;
  }

  public List<Property> getProperties() {
    return properties;
  }

  public PeripheralJobStatusMessage setProperties(List<Property> properties) {
    this.properties = properties;
    return this;
  }
}
