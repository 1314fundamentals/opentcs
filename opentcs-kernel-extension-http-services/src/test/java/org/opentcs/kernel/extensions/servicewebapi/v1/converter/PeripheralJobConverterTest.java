// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.time.Instant;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.opentcs.data.model.Location;
import org.opentcs.data.model.LocationType;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.data.peripherals.PeripheralOperation;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.GetPeripheralJobResponseTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents.PeripheralJobStatusMessage;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel.PeripheralJobStateTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel.PeripheralOperationTO;

/**
 * Tests for {@link PeripheralJobConverter}.
 */
public class PeripheralJobConverterTest {
  private final PeripheralJobConverter peripheralJobConverter = new PeripheralJobConverter(
      new PeripheralOperationConverter()
  );

  @Test
  void checkToPeripheralJobStatusMessage() {
    PeripheralJob job = new PeripheralJob(
        "some-job",
        "some-token",
        new PeripheralOperation(
            new Location("L1", new LocationType("LT1").getReference()).getReference(),
            "some-operation",
            PeripheralOperation.ExecutionTrigger.AFTER_ALLOCATION,
            true
        )
    )
        .withRelatedVehicle(new Vehicle("V1").getReference())
        .withRelatedTransportOrder(new TransportOrder("T1", List.of()).getReference())
        .withState(PeripheralJob.State.TO_BE_PROCESSED)
        .withCreationTime(Instant.EPOCH)
        .withFinishedTime(Instant.EPOCH)
        .withProperty("some-prop-key", "some-prop-value");

    PeripheralJobStatusMessage statusMessage = peripheralJobConverter.toPeripheralJobStatusMessage(
        job, 1L, Instant.EPOCH
    );
    assertThat(statusMessage.getName(), is("some-job"));
    assertThat(statusMessage.getReservationToken(), is("some-token"));
    assertThat(statusMessage.getPeripheralOperation().getOperation(), is("some-operation"));
    assertThat(statusMessage.getPeripheralOperation().getLocationName(), is("L1"));
    assertThat(
        statusMessage.getPeripheralOperation().getExecutionTrigger(), is(
            PeripheralOperationTO.ExecutionTrigger.AFTER_ALLOCATION
        )
    );
    assertThat(statusMessage.getPeripheralOperation().isCompletionRequired(), is(true));
    assertThat(
        statusMessage.getState(), is(PeripheralJobStateTO.TO_BE_PROCESSED)
    );
    assertThat(statusMessage.getRelatedVehicle(), is("V1"));
    assertThat(statusMessage.getRelatedTransportOrder(), is("T1"));
    assertThat(statusMessage.getCreationTime(), is(Instant.EPOCH));
    assertThat(statusMessage.getFinishedTime(), is(Instant.EPOCH));
    assertThat(statusMessage.getProperties().size(), is(1));
    assertThat(statusMessage.getProperties().getFirst().getKey(), is("some-prop-key"));
    assertThat(statusMessage.getProperties().getFirst().getValue(), is("some-prop-value"));
  }

  @Test
  void checkToGetPeripheralJobResponse() {
    PeripheralJob job = new PeripheralJob(
        "some-job",
        "some-token",
        new PeripheralOperation(
            new Location("L1", new LocationType("LT1").getReference()).getReference(),
            "some-operation",
            PeripheralOperation.ExecutionTrigger.AFTER_ALLOCATION,
            true
        )
    )
        .withRelatedVehicle(new Vehicle("V1").getReference())
        .withRelatedTransportOrder(new TransportOrder("T1", List.of()).getReference())
        .withState(PeripheralJob.State.TO_BE_PROCESSED)
        .withCreationTime(Instant.EPOCH)
        .withFinishedTime(Instant.EPOCH)
        .withProperty("some-prop-key", "some-prop-value");

    GetPeripheralJobResponseTO response = peripheralJobConverter.toGetPeripheralJobResponseTO(job);
    assertThat(response.getName(), is("some-job"));
    assertThat(response.getReservationToken(), is("some-token"));
    assertThat(response.getPeripheralOperation().getOperation(), is("some-operation"));
    assertThat(response.getPeripheralOperation().getLocationName(), is("L1"));
    assertThat(
        response.getPeripheralOperation().getExecutionTrigger(), is(
            PeripheralOperationTO.ExecutionTrigger.AFTER_ALLOCATION
        )
    );
    assertThat(response.getPeripheralOperation().isCompletionRequired(), is(true));
    assertThat(response.getState(), is(PeripheralJobStateTO.TO_BE_PROCESSED));
    assertThat(response.getRelatedVehicle(), is("V1"));
    assertThat(response.getRelatedTransportOrder(), is("T1"));
    assertThat(response.getCreationTime(), is(Instant.EPOCH));
    assertThat(response.getFinishedTime(), is(Instant.EPOCH));
    assertThat(response.getProperties().size(), is(1));
    assertThat(response.getProperties().getFirst().getKey(), is("some-prop-key"));
    assertThat(response.getProperties().getFirst().getValue(), is("some-prop-value"));
  }
}
