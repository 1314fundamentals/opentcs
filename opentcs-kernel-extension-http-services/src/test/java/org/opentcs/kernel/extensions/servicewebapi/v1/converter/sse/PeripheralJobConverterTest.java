// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter.sse;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.assertj.core.groups.Tuple;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.opentcs.data.ObjectHistory;
import org.opentcs.data.model.Location;
import org.opentcs.data.model.LocationType;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.data.peripherals.PeripheralOperation;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.TimestampScrubber;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.ObjectHistoryTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.PeripheralJobEventTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.PeripheralJobEventTO.PeripheralJobTO.PeripheralOperationTO.ExecutionTriggerTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.PeripheralJobEventTO.PeripheralJobTO.StateTO;

/**
 * Tests for {@link PeripheralJobConverter}.
 */
class PeripheralJobConverterTest {

  private JsonBinder jsonBinder;
  private PeripheralJobConverter converter;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
    converter = new PeripheralJobConverter();
  }

  @Test
  void convert() {
    PeripheralJob peripheralJob = new PeripheralJob(
        "job-1",
        "reservation-token-1",
        new PeripheralOperation(
            new Location(
                "location-1",
                new LocationType("location-type-1").getReference()
            ).getReference(),
            "operation-1",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    )
        .withProperties(Map.of("key-1", "value-1", "key-2", "value-2"))
        .withHistoryEntry(
            new ObjectHistory.Entry(Instant.EPOCH, "event-code-1", List.of("supplement-1"))
        )
        .withRelatedVehicle(new Vehicle("vehicle-1").getReference())
        .withRelatedTransportOrder(new TransportOrder("order-1", List.of()).getReference())
        .withState(PeripheralJob.State.FINISHED)
        .withCreationTime(Instant.EPOCH)
        .withFinishedTime(Instant.EPOCH);

    PeripheralJobEventTO.PeripheralJobTO result = converter.convert(peripheralJob);

    Approvals.verify(
        jsonBinder.toJson(result),
        // Scrub timestamps in the result's JSON representation as some of them may be generated
        // during test runtime, making it impossible to use ApprovalTests.
        new Options(new TimestampScrubber())
    );
  }

  /**
   * Ensures that timestamps are converted correctly. (Testing this is necessary, since timestamps
   * are scrubbed in the ApprovalTest and therefore their conversion is not covered there.)
   */
  @Test
  void convertsTimestampsCorrectly() {
    Instant time = Instant.parse("2025-07-23T12:03:00Z");
    PeripheralJob peripheralJob = new PeripheralJob(
        "job-1",
        "reservation-token-1",
        new PeripheralOperation(
            new Location(
                "location-1",
                new LocationType("location-type-1").getReference()
            ).getReference(),
            "operation-1",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    )
        .withCreationTime(time)
        .withFinishedTime(time.plus(1, ChronoUnit.HOURS))
        .withHistoryEntry(
            new ObjectHistory.Entry(
                time.plus(2, ChronoUnit.HOURS),
                "event-code-1",
                List.of("supplement-1")
            )
        );

    PeripheralJobEventTO.PeripheralJobTO result = converter.convert(peripheralJob);

    assertThat(result.getCreationTime()).isEqualTo(time);
    assertThat(result.getFinishedTime()).isEqualTo(time.plus(1, ChronoUnit.HOURS));
    assertThat(result.getHistory().getEntries())
        .extracting(
            ObjectHistoryTO.ObjectHistoryEntryTO::getTimestamp,
            ObjectHistoryTO.ObjectHistoryEntryTO::getEventCode,
            ObjectHistoryTO.ObjectHistoryEntryTO::getSupplements
        )
        .contains(
            Tuple.tuple(
                time.plus(2, ChronoUnit.HOURS),
                "event-code-1",
                List.of("supplement-1")
            )
        );
  }

  @ParameterizedTest
  @EnumSource(PeripheralJob.State.class)
  void convertsAllStates(PeripheralJob.State state) {
    PeripheralJob job = new PeripheralJob(
        "job-1",
        "reservation-token-1",
        new PeripheralOperation(
            new Location(
                "location-1",
                new LocationType("location-type-1").getReference()
            ).getReference(),
            "operation-1",
            PeripheralOperation.ExecutionTrigger.AFTER_MOVEMENT,
            true
        )
    )
        .withState(state);

    PeripheralJobEventTO.PeripheralJobTO result = converter.convert(job);

    StateTO expectedState = switch (state) {
      case TO_BE_PROCESSED -> StateTO.TO_BE_PROCESSED;
      case BEING_PROCESSED -> StateTO.BEING_PROCESSED;
      case FINISHED -> StateTO.FINISHED;
      case FAILED -> StateTO.FAILED;
    };
    assertThat(result.getState()).isEqualTo(expectedState);
  }

  @ParameterizedTest
  @EnumSource(PeripheralOperation.ExecutionTrigger.class)
  void convertsAllExecutionTriggers(PeripheralOperation.ExecutionTrigger executionTrigger) {
    PeripheralJob job = new PeripheralJob(
        "job-1",
        "reservation-token-1",
        new PeripheralOperation(
            new Location(
                "location-1",
                new LocationType("location-type-1").getReference()
            ).getReference(),
            "operation-1",
            executionTrigger,
            true
        )
    );

    PeripheralJobEventTO.PeripheralJobTO result = converter.convert(job);

    ExecutionTriggerTO expectedExecutionTrigger = switch (executionTrigger) {
      case IMMEDIATE -> ExecutionTriggerTO.IMMEDIATE;
      case AFTER_ALLOCATION -> ExecutionTriggerTO.AFTER_ALLOCATION;
      case AFTER_MOVEMENT -> ExecutionTriggerTO.AFTER_MOVEMENT;
    };
    assertThat(result.getPeripheralOperation().getExecutionTrigger())
        .isEqualTo(expectedExecutionTrigger);
  }
}
