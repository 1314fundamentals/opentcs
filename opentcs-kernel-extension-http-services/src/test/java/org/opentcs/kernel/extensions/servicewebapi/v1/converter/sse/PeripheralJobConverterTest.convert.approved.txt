{
  "name" : "job-1",
  "properties" : {
    "key-1" : "value-1",
    "key-2" : "value-2"
  },
  "history" : {
    "entries" : [ {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:jobCreated",
      "supplements" : [ ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "event-code-1",
      "supplements" : [ "supplement-1" ]
    } ]
  },
  "reservationToken" : "reservation-token-1",
  "relatedVehicle" : "vehicle-1",
  "relatedTransportOrder" : "order-1",
  "peripheralOperation" : {
    "location" : "location-1",
    "operation" : "operation-1",
    "executionTrigger" : "AFTER_MOVEMENT",
    "completionRequired" : true
  },
  "state" : "FINISHED",
  "creationTime" : "[Timestamp]",
  "finishedTime" : "[Timestamp]"
}